include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -7176,11 +7176,24 @@ namespace VectorTools
               Assert (n_components >= dim,
                       ExcMessage ("You can only ask for the Hdiv norm for a finite element "
                                   "with at least 'dim' components. In that case, this function "
-                                  "will take the divergence of the first 'dim' components."));
+                                  "will find the index of the first non-zero weight and take "
+                                  "the divergence of the 'dim' components that follow it."));
+
+              unsigned int idx = 0;
+              if (weight!=nullptr)
+                {
+                  Vector<double> tmp(n_components);
+                  weight->vector_value(Point<spacedim>(), tmp);
+
+                  for (; idx<n_components; ++idx)
+                    if (fabs(tmp[idx]) > 1.e-12)
+                      break;
+                }
+
               Number sum = 0;
               // take the trace of the derivatives scaled by the weight and square it
-              for (unsigned int k=0; k<dim; ++k)
-                sum += data.psi_grads[q][k][k] * std::sqrt(data.weight_vectors[q](k));
+              for (unsigned int k=idx; k<idx+dim; ++k)
+                sum += data.psi_grads[q][k][k-idx] * std::sqrt(data.weight_vectors[q](k));
```

Comment:
We should add an assertion before this so guarantee that `data.psi_grads[q][k]` exists.

Reply:
@drwells Doesn't `IDScratchData` method `resize_vectors` guarantee this? It sets `psi_grads` to have the correct sizes and correct tensor dimension. Could you please elaborate, what should this assertion check?

Reply:
I should have been clearer. We currently check that `n_components >= dim`. We should move this check to after the point where we calculate a (possibly nonzero) value for `idx` and check that `n_components >= dim + idx` instead.